server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java

```diff
@@ -0,0 +1,384 @@
+
+        if (values.isEmpty()) {
+            return null;
+        }
+
+        return values.get(0);
+    }
+
+    @Override
+    public boolean isReadOnlyAttribute(String key) {
+        if (readOnlyInternalPredicate.test(key)) {
+            return true;
+        }
+
+        List<AttributeValidator> validators = this.validators.getOrDefault(key, Collections.emptyList());
+
+        if (validators.isEmpty()) {
+            // the attribute is configurable so we try to validate global validations
+            validators = this.validators.getOrDefault(READ_ONLY_ATTRIBUTE_KEY, Collections.emptyList());
+
+            for (AttributeValidator validator : validators) {
+                if (!validator.getValidator().validate(new Entry<String, List<String>>() {
+                    @Override
+                    public String getKey() {
+                        return key;
+                    }
+
+                    @Override
+                    public List<String> getValue() {
+                        return Collections.emptyList();
+                    }
+
+                    @Override
+                    public List<String> setValue(List<String> value) {
+                        return Collections.emptyList();
+                    }
+                }, user)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean validate(String name, BiConsumer<Entry<String, List<String>>, String>... listeners) {
+        Entry<String, List<String>> attribute = new SimpleEntry<String, List<String>>(name, null) {
+            @Override
+            public List<String> getValue() {
+                List<String> values = get(name);
+
+                if (values == null) {
+                    return EMPTY_VALUE;
+                }
+
+                return values;
+            }
+        };
+
+        // firstly, validate core attributes like username
+        // TODO: we should be able to move most of the validation from legacy validators to this method
+        if (!validateCoreAttributes(attribute, listeners)) {
+            return false;
+        }
+
+        List<AttributeValidator> validators = this.validators.getOrDefault(attribute.getKey(), new ArrayList<>());
+
+        validators.addAll(this.validators.getOrDefault(READ_ONLY_ATTRIBUTE_KEY, Collections.emptyList()));
+
+        boolean success = true;
+
+        for (AttributeValidator validator : validators) {
+            if (!validator.getValidator().validate(attribute, user)) {
+                success = false;
+                if (listeners != null) {
+                    triggerValidationListeners(attribute, validator.getMessage(), listeners);
+                }
+            }
+        }
+
+        return success;
+    }
+
+    private boolean validateCoreAttributes(Entry<String, List<String>> attribute, BiConsumer<Entry<String, List<String>>, String>[] listeners) {
```

Comment:
make this protected to allow customization please

Reply:
This method no longer exists. This class should rely on whatever validations were set by the provider (including for core attributes).